client/src/language/codeActionProvider.ts

```diff
@@ -0,0 +1,101 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import vscode from 'vscode'
+import { requestsManager } from './RequestManager'
+import { embeddedLanguageDocsManager } from './EmbeddedLanguageDocsManager'
+import { getEmbeddedLanguageDocRange, getIndentationOnLine, getOriginalDocPosition } from './utils'
+import { logger } from '../lib/src/utils/OutputLogger'
+import { type Range } from 'vscode-languageclient'
+
+export class BitbakeCodeActionProvider implements vscode.CodeActionProvider {
+  async provideCodeActions (document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<vscode.CodeAction[]> {
+    const diagnostics = context.diagnostics
+    const actions = await Promise.all(
+      diagnostics.map(async (diagnostic) => await buildActionFromDiagnostic(document, diagnostic))
+    ).then((results) => results.flat())
+    return actions
+  }
+}
+
+const buildActionFromDiagnostic = async (document: vscode.TextDocument, diagnostic: vscode.Diagnostic): Promise<vscode.CodeAction[]> => {
+  const originalRange = diagnostic.range
+  const embeddedLanguageType = await requestsManager.getEmbeddedLanguageTypeOnPosition(document.uri.toString(), originalRange.start)
+  if (embeddedLanguageType === undefined || embeddedLanguageType === null) {
+    return []
+  }
+  const embeddedLanguageDocInfos = embeddedLanguageDocsManager.getEmbeddedLanguageDocInfos(document.uri.toString(), embeddedLanguageType)
+  if (embeddedLanguageDocInfos === undefined || embeddedLanguageDocInfos === null) {
+    return []
+  }
+  const embeddedLanguageTextDocument = await vscode.workspace.openTextDocument(embeddedLanguageDocInfos.uri)
```

Comment:
Did you test if this did not deteriorate our virtual files opening/flooding issues?

Reply:
The "opening/flooding issues" have been stabilized. It was due to "updateDiagnostics" reopening files that had been closed by VS Code itself. Opening the file again would generate new diagnostics (identical to the previous ones), that would trigger "onDidChangeDiagnostics", which would call "updateDiagnostics" once more.

Here, calling `openTextDocument` has no way to have the side effect to call again `provideCodeActions`. Furthermore, the fact the user can click on the quick fixes means the "embedded language document" is currently "opened" by VS Code, since the user can see its diagnostics. So `openTextDocument` will simply "return early", as stated in the doc.